### PR TITLE
Remove ORDER BY applied to singleton queries

### LINKED ISSUE #34831: Remove `ORDER BY` applied to singleton queries
When a (sub)query results in 0 or 1 elements (First(), Last(), FirstOrDefault(), LastOrDefault(), Take(1), ...), `OrderBy`/`ThenBy` has no effect and can be ignored. This can avoid a pushdown and in general more complex queries.

An example that could be improved:

```csharp
// @nuget: Microsoft.EntityFrameworkCore.Sqlite -Version 8.0.8

using System;
using System.Data;
using System.Linq;
using Microsoft.EntityFrameworkCore;

using var db = new BloggingContext();

var query = (from blogs in db.Blogs
	select (from posts in blogs.Posts select posts).OrderBy(e => e.SortKey).Take(1).OrderBy(e => e.Id).Single().Id);

Console.WriteLine(query.ToQueryString());

public class BloggingContext : DbContext
{
    public DbSet<Blog> Blogs { get; set; }
    public DbSet<Post> Posts { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder options)
        => options
            .LogTo(Console.WriteLine, Microsoft.Extensions.Logging.LogLevel.Information)
            .UseSqlite($"Data Source=test.db");
}

public class Blog
{
    public int Id { get; set; }
    public Post[] Posts { get; set; }
}

public class Post
{
    public int Id { get; set; }
    public Blog Blog { get; set; }
    public int SortKey { get; set; }
}
```

In this case the SQL query is
```sql
SELECT (
    SELECT "t"."Id"
    FROM (
        SELECT "p"."Id", "p"."BlogId", "p"."SortKey"
        FROM "Posts" AS "p"
        WHERE "b"."Id" = "p"."BlogId"
        ORDER BY "p"."SortKey"
        LIMIT 1
    ) AS "t"
    ORDER BY "t"."Id"
    LIMIT 1)
FROM "Blogs" AS "b"
```
but it could simply be
```sql
SELECT (
        SELECT "p"."Id"
        FROM "Posts" AS "p"
        WHERE "b"."Id" = "p"."BlogId"
        ORDER BY "p"."SortKey"
        LIMIT 1)
FROM "Blogs" AS "b"
```

This can combine with https://github.com/dotnet/efcore/issues/34482 to further simplify queries.

### PR DESCRIPTION
**Pull request by ranma42**
Fixes #34831.

### CODE REVIEW

**Comment by ranma42** [comment]
I added a test and based this on top of #35384, so that it actually performs the desired simplifications (as visible in `Distinct_take_without_orderby`).

**Comment by roji** [comment]
@ranma42 thanks as always for the work and sorry it took so long to review. I'm coming out of a very intensive work period working on other stuff and will try to review PRs etc.

This change seems fine to me, though IIUC the kind of queries this optimize are "silly", if written by an end user; in other words, a LINQ query (EF or otherwise) containing `Take(1).OrderBy(...)` doesn't make any sense, as the OrderBy can always be removed (a .NET analyzer could in theory be written which does that, if we knew that the OrderBy had no side effects). 

We generally avoid going too much into this kind of optimization, where EF tries to catch badly-written queries and improve their translations; it's better for users to write better LINQ queries. It's true that in some cases such optimizations can kick in as a result of processing we ourselves do in the query pipeline (and not because of "silly" input queries); that's indeed valuable, though is that the case here?

Bottom line: I'm totally fine merging this as the change adds minimal complexity, but I'm just asking in case I'm missing something (or you want to discuss). Let me know what you think.

**Comment by roji** [comment]
@ranma42 any comment on the above?

**Comment by ranma42** [comment]
oops, sorry for the delayed reply!

I agree that these queries are unlikely in a normal codebase: `.Take(N).OrderBy()` is unusual and even more so the special case N=1, which is obviously pointless, so it can only happen because of a mistake or multiple parts that compose a query without tracking its cardinality.

OTOH this is a simple check that brings the expressions a little more towards a "normalized" form.
Conceptually, this is a generalization of #34482 (DISTINCT over singletons is a no-op, just like ORDER BY)

> We generally avoid going too much into this kind of optimization, where EF tries to catch badly-written queries and improve their translations; it's better for users to write better LINQ queries. It's true that in some cases such optimizations can kick in as a result of processing we ourselves do in the query pipeline (and not because of "silly" input queries); that's indeed valuable, though is that the case here?

It was not designed for this, although given the code that ended up interacting with (see both the changes and the interaction with #34097) it might indeed trigger in some split queries 🤔 I will try and have a look (I am thinking about something like `db.Blogs.Include(x => x.Posts.OrderBy(y => y.SortKey).Take(1))`, which is definitely non-standard, but somewhat reasonable and expressed in the most natural form).

I will investigate the split query case and report back; in any case, either decision (merge or close) works for me 👍
